### ISSUE TITLE
"The requested version does not exist" for yanked crates

### ISSUE THREAD

**Issue by WalterSmuts** [rust-lang/docs.rs#1822]
For situations where all the versions of a crate has been yanked you get the following error:
```
The requested version does not exist
```
when visiting the latest url: 
* `https://docs.rs/$YANKED_CRATE_NAME`.
* `https://docs.rs/$YANKED_CRATE_NAME/latest/`.
* `https://docs.rs/$YANKED_CRATE_NAME/latest/$YANKED_CRATE_NAME/index.html`.

Generally when crate authors would share links to the documentation for their crate they would share a link to the generic "latest" page. This means all of those links are dead even though the author may have pushed a final version explaining an alternative crate to use, before yanking that version too.

Real life example:
https://docs.rs/constfft/latest/constfft/
https://docs.rs/constfft/0.2.4/constfft/index.html

Suggestion: redirect to the most docs, even when all the versions are yanked but add a banner that no un-yanked versions exist anymore.

**Comment by jyn514** [comment]
Take a look at https://github.com/rust-lang/docs.rs/blob/master/src/web/routes.rs to figure out what serves those routes - I think something is handling the output of `match_version` incorrectly.

**Comment by RushilMehra** [comment]
I'll give this issue a stab

**Comment by Nemo157** [comment]
> Take a look at https://github.com/rust-lang/docs.rs/blob/a074c8a3ad66857bae654454e1eee507bb76422d/src/web/routes.rs to figure out what serves those routes - I think something is handling the output of `match_version` incorrectly.

This being a 404 was an explicit design choice from #322

> * A change to the version resolution code to look up the list of releases of the crate that are marked as yanked in our database, and avoid resolving to yanked releases unless explicitly asked for.
>   * (Side note: if all of a crate's versions have been yanked, this will make it appear to not exist if you go to docs.rs/crate. Is this behavior we want? I can make it check whether all versions have been yanked before filtering them out, if desired.)

**Comment by jyn514** [comment]
I think if all versions of a crate have been yanked, it makes sense to show the latest yanked version, rather than a 404.

**Comment by Nemo157** [comment]
Yup, I think it could make sense to change, just clarifying where the behavior came from (and there are explicit tests for the current behavior that will need updating).

**Comment by RushilMehra** [comment]
That makes sense, if everyone agrees then I'll work on improving the messaging on the page

**Comment by WalterSmuts** [comment]
> After just fully yanking one of my own old unused crates I actually like the current design. Yanking all versions without publishing a new unyanked one is a really strong "I don't want anyone using this" signal. I was thinking we probably should also remove fully yanked crates from authors release pages too.

I agree we should heavily signal that the crate should not be used. In fact, it CANNOT be used unless you have it in your `Cargo.lock` file already. My situation for submitting this issue was that there was an alternate that was appropriate and I as a crate author would like to direct folks to use alternate. My only way of doing this is to update the `README` to point the user to the alternate. The current behaviour doesn't display any `README` at all.

> This means all of those links are dead even though the author may have pushed a final version explaining an alternative crate to use, before yanking that version too

This is the situation I'm trying to address.
